### PR TITLE
chartsync: support Git HTTPS credentials using secret

### PR DESCRIPTION
**Pull request by richardcase**
Started to add https basic authentication to the git chart source. the
username and password can be set as a static value or from a secret.
For example:

```yaml
spec:
  releaseName: my-app
  chart:
    git: https://github.com/someonesomerepo.git
    ref: master
    path: charts/my-app
    secretRef:
      name: git-credentials

```

Still a work in progress

Relates to: #94

### CODE REVIEW

**Comment by stefanprodan** [comment]
@richardcase I think this should be paused until we decide how to handle git/helm repos outside the HelmRelease spec. See https://github.com/fluxcd/helm-operator/issues/142

**Comment by richardcase** [comment]
> @richardcase I think this should be paused until we decide how to handle git/helm repos outside the HelmRelease spec. See #142

@stefanprodan  - Sure. Although i really need this functionality for a specific scenario i'm working on. I'll work from a fork. If you need any help with #142 let me know

**Comment by stefanprodan** [comment]
If you need this then please make it more compact like in my proposal:

```yaml
spec:
  releaseName: my-app
  chart:
    git: https://github.com/someonesomerepo.git
    ref: master
    path: charts/my-app
    secretRef: #optional
      name: git-basic-auth #required
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-basic-auth
type: Opaque
data:
  username: <BASE64> #required
  password: <BASE64> #required
```

**Comment by richardcase** [comment]
> If you need this then please make it more compact like in my proposal

Sure i can do that. I will make that change now

**Comment by hiddeco** [comment]
In addition to all of the above, instead of merging in `master` it is better for reviewers and Git history clutter to rebase on top of `master`.

**Comment by richardcase** [comment]
> In addition to all of the above, instead of merging in `master` it is better for reviewers and Git history clutter to rebase on top of `master`.

Understood, that was a mess up my side....a case of PEBKAC

**Comment by richardcase** [comment]
@stefanprodan @hiddeco  - i've made the change to simplify the specifiying of the creds as suggested. So as suggested you have:

```yaml
spec:
  releaseName: my-release
  chart:
    git: https://github.com/someorg/somerepo.git
    ref: master
    path: charts/my-chart
    secretRef:
      name: git-pwd
```
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-pwd
type: Opaque
data:
  password: supersecret
  username: joeblogs
```

File: cmd/helm-operator/main.go [hunk 1]
```diff
@@ -23,8 +23,8 @@ import (
 	clientset "github.com/fluxcd/helm-operator/pkg/client/clientset/versioned"
 	ifinformers "github.com/fluxcd/helm-operator/pkg/client/informers/externalversions"
 	"github.com/fluxcd/helm-operator/pkg/helm"
-	"github.com/fluxcd/helm-operator/pkg/helm/v2"
-	"github.com/fluxcd/helm-operator/pkg/helm/v3"
+	v2 "github.com/fluxcd/helm-operator/pkg/helm/v2"
+	v3 "github.com/fluxcd/helm-operator/pkg/helm/v3"
```
Comment (hiddeco) [inline cmd/helm-operator/main.go]: The alias here is obsolete, and is probably caused by some enabled linter.

File: pkg/apis/helm.fluxcd.io/v1/types.go [hunk 2]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/ghodss/yaml"
 
 	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (hiddeco) [inline pkg/apis/helm.fluxcd.io/v1/types.go]: These are duplicates (`corev1` should be kept).

File: pkg/chartsync/git.go [hunk 3]
```diff
@@ -3,16 +3,22 @@ package chartsync
...
 	"github.com/go-kit/kit/log"
 
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
 	"k8s.io/client-go/tools/cache"
 
-	"github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
+	v1 "github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
```
Comment (hiddeco) [inline pkg/chartsync/git.go]: Same as the above.

File: pkg/chartsync/git.go [hunk 4]
```diff
@@ -3,16 +3,22 @@ package chartsync
...
+	"net/url"
+	"strings"
 	"sync"
 	"time"
 
 	"github.com/fluxcd/flux/pkg/git"
 	"github.com/go-kit/kit/log"
 
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (hiddeco) [inline pkg/chartsync/git.go:16]: :+1:

File: pkg/chartsync/git.go [hunk 5]
```diff
@@ -281,12 +291,22 @@ func (c *GitChartSync) sync(hr *v1.HelmRelease, mirrorName string, repo *git.Rep
 // maybeMirror requests a new mirror for the given remote. The return value
 // indicates whether the repo was already present (`true` if so,
 // `false` otherwise).
-func (c *GitChartSync) maybeMirror(mirrorName string, remote string) bool {
+func (c *GitChartSync) maybeMirror(mirrorName string, source *v1.GitChartSource, namespace string) bool {
+	gitURL := source.GitURL
+	if source.SecretRef != nil {
```
Comment (hiddeco) [inline pkg/chartsync/git.go]: As you are doing the same `nil` check in `addAuthForHTTPS`, I would reduce this to

```go
gitURL := source.GitURL
var err error
if gitURL, err = c.addAuthForHTTPS(source.GitURL, source.SecretRef, namespace); err != nil {
  c.logger.Log("error", GitAuthError{err}.Error())
  return false
}
```

**Review by hiddeco — COMMENTED** [review]
Besides the in-code comments, I think the `mirrorName` function needs to be made smarter as the credentials from other `HelmRelease` resources (and namespaces) will now leak by configuring a `source.GitURL` (without credentials) that is already in use by another `HelmRelease` (with credentials).

You may get away with just adding the namespace, if I am not mistaken this is the same approach Flux uses for image pull secrets.

File: cmd/helm-operator/main.go (hunk 1 above)
Comment (richardcase) [reply cmd/helm-operator/main.go]: Changed this.

File: pkg/chartsync/git.go (hunk 3 above)
Comment (richardcase) [reply pkg/chartsync/git.go]: @hiddeco - every time i removed the v1 alias it adds it back so this hasn't been resolved i'm afraid

File: pkg/apis/helm.fluxcd.io/v1/types.go (hunk 2 above)
Comment (richardcase) [reply pkg/apis/helm.fluxcd.io/v1/types.go]: Changed

File: pkg/chartsync/git.go (hunk 5 above)
Comment (richardcase) [reply pkg/chartsync/git.go]: A good suggestion, changed.

**Comment by richardcase** [comment]
> You may get away with just adding the namespace, if I am not mistaken this is the same approach Flux uses for image pull secrets.

@hiddeco - i checked and flux does use the namespace. I changed the `mirrorName` to use a composite of namespace, auth secret name, git url. Do you think this is ok?

File: pkg/chartsync/git.go (hunk 3 above)
Comment (hiddeco) [reply pkg/chartsync/git.go]: I will remove it for you before the merge, no worries.

File: pkg/chartsync/git.go [hunk 6]
```diff
@@ -314,7 +332,60 @@ func (c *GitChartSync) helmReleasesForMirror(mirror string) ([]*v1.HelmRelease,
 // per namespace, per auth.
 func mirrorName(hr *v1.HelmRelease) string {
 	if hr != nil && hr.Spec.GitChartSource != nil {
-		return hr.Spec.GitURL
+		secretName := "noauth"
```
Comment (hiddeco) [inline pkg/chartsync/git.go]: I think this approach is ideologically correct when secrets are being used, but this also means we will do a lot more (duplicate) polling for e.g. releases without HTTPS authentication. My propose would be to change it to something like this:

```go
if hr != nil && hr.Spec.GitChartSource != nil {
	if hr.Spec.GitChartSource.SecretRef == nil {
		return hr.Spec.GitURL
	}
	return fmt.Sprintf("%s/%s/%s", hr.GetNamespace(), hr.Spec.GitChartSource.SecretRef.Name, hr.Spec.GitURL)
}
return ""
```

In addition, the `TODO(michael)` comment can also be removed.

**Comment by hiddeco** [comment]
@richardcase thanks a lot :1st_place_medal: 

@stefanprodan can you take a look and see if this matches your expectations for the issue you mentioned? Feature wise it looks fine and mergeable to me (after I get rid of the persistent alias caused by the linter).

**Comment by stefanprodan** [comment]
@hiddeco it looks Ok to me, it needs a rebase with master are you going to do that when fixing the aliases?

**Comment by hiddeco** [comment]
@stefanprodan yep, will do the rebase and some minor fixes that triggered my OCD :+1:

**Review by hiddeco — APPROVED** [review]
Rebased and tidied up, awesome contribution @richardcase :raised_hands: :cake:

File: pkg/chartsync/git.go [hunk 7]
```diff
@@ -284,12 +292,20 @@ func (c *GitChartSync) sync(hr *v1.HelmRelease, mirrorName string, repo *git.Rep
...
+		return false
+	}
+
 	ok := c.mirrors.Mirror(
-		mirrorName, git.Remote{URL: remote}, git.Timeout(c.config.GitTimeout),
+		mirrorName, git.Remote{URL: gitURL}, git.Timeout(c.config.GitTimeout),
 		git.PollInterval(c.config.GitPollInterval), git.ReadOnly)
 	if !ok {
-		c.logger.Log("info", "started mirroring new remote", "remote", remote, "mirror", mirrorName)
+		c.logger.Log("info", "started mirroring new remote", "remote", source.GitURL, "mirror", mirrorName)
```
Comment (stefanprodan) [inline pkg/chartsync/git.go:308]: Do we log the credentials here?

File: pkg/chartsync/git.go [hunk 8]
```diff
@@ -78,11 +85,12 @@ func (c sourceRef) forHelmRelease(hr *v1.HelmRelease) bool {
 }
 
 func NewGitChartSync(logger log.Logger,
-	lister lister.HelmReleaseLister, cfg GitConfig, queue ReleaseQueue) *GitChartSync {
+	coreV1Client corev1client.CoreV1Interface, lister lister.HelmReleaseLister, cfg GitConfig, queue ReleaseQueue) *GitChartSync {
 
 	return &GitChartSync{
 		logger:             logger,
 		config:             cfg,
+		coreV1Client: 		coreV1Client,
```
Comment (stefanprodan) [inline pkg/chartsync/git.go]: Please run `go fmt`.

File: pkg/chartsync/git.go (hunk 7 above)
Comment (hiddeco) [reply pkg/chartsync/git.go:308]: No, `source.GitURL` does not include the credentials but is the 1:1 copy from the `HelmRelease` definition.

Comment (stefanprodan) [reply pkg/chartsync/git.go:308]: Great!

**Comment by stefanprodan** [comment]
In the future we should consider adding support for cross-namespace secrets like we have done in https://github.com/fluxcd/helm-operator/pull/219

**Review by stefanprodan — APPROVED** [review]
LGTM

**Pull request merged**